In [1]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns




In [2]:
df = pd.read_csv('house_prices.csv')

In [3]:
df


area_type   availability                  location  \
0      Super built-up  Area         19-Dec  Electronic City Phase II   
1                Plot  Area  Ready To Move          Chikka Tirupathi   
2            Built-up  Area  Ready To Move               Uttarahalli   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli   
4      Super built-up  Area  Ready To Move                  Kothanur   
...                     ...            ...                       ...   
13315        Built-up  Area  Ready To Move                Whitefield   
13316  Super built-up  Area  Ready To Move             Richards Town   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar   
13318  Super built-up  Area         18-Jun           Padmanabhanagar   
13319  Super built-up  Area  Ready To Move              Doddathoguru   

            size  society total_sqft  bath  balcony   price  
0          2 BHK  Coomee        1056   2.0      1.0   39.07  
1      4 Bedroom  Theanmp       2600   5.0      3.0  120.00  
2          3 BHK      NaN       1440   2.0      3.0   62.00  
3          3 BHK  Soiewre       1521   3.0      1.0   95.00  
4          2 BHK      NaN       1200   2.0      1.0   51.00  
...          ...      ...        ...   ...      ...     ...  
13315  5 Bedroom  ArsiaEx       3453   4.0      0.0  231.00  
13316      4 BHK      NaN       3600   5.0      NaN  400.00  
13317      2 BHK  Mahla T       1141   2.0      1.0   60.00  
13318      4 BHK  SollyCl       4689   4.0      1.0  488.00  
13319      1 BHK      NaN        550   1.0      1.0   17.00  

[13320 rows x 9 columns]

In [4]:
df2  = df.drop(['area_type','society','availability' ,'balcony'] , axis=1)

In [5]:
df2

location       size total_sqft  bath   price
0      Electronic City Phase II      2 BHK       1056   2.0   39.07
1              Chikka Tirupathi  4 Bedroom       2600   5.0  120.00
2                   Uttarahalli      3 BHK       1440   2.0   62.00
3            Lingadheeranahalli      3 BHK       1521   3.0   95.00
4                      Kothanur      2 BHK       1200   2.0   51.00
...                         ...        ...        ...   ...     ...
13315                Whitefield  5 Bedroom       3453   4.0  231.00
13316             Richards Town      4 BHK       3600   5.0  400.00
13317     Raja Rajeshwari Nagar      2 BHK       1141   2.0   60.00
13318           Padmanabhanagar      4 BHK       4689   4.0  488.00
13319              Doddathoguru      1 BHK        550   1.0   17.00

[13320 rows x 5 columns]

In [6]:
df2.isnull().sum()

location       1
size          16
total_sqft     0
bath          73
price          0
dtype: int64

In [7]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13319 non-null  object 
 1   size        13304 non-null  object 
 2   total_sqft  13320 non-null  object 
 3   bath        13247 non-null  float64
 4   price       13320 non-null  float64
dtypes: float64(2), object(3)
memory usage: 520.4+ KB


out off 13320 entries only few are null so can drop them

In [8]:
df3 = df2.dropna()

In [9]:
df3.isnull().sum()

location      0
size          0
total_sqft    0
bath          0
price         0
dtype: int64

In [10]:
df3 = df3.copy()
# used this cause df3 is just a view , for working on a dataframe need to make a dataframe by using .copy()
df3['bhk'] = df3['size'].apply(lambda x : x.split(' ')[0])

In [11]:
df3

location       size total_sqft  bath   price bhk
0      Electronic City Phase II      2 BHK       1056   2.0   39.07   2
1              Chikka Tirupathi  4 Bedroom       2600   5.0  120.00   4
2                   Uttarahalli      3 BHK       1440   2.0   62.00   3
3            Lingadheeranahalli      3 BHK       1521   3.0   95.00   3
4                      Kothanur      2 BHK       1200   2.0   51.00   2
...                         ...        ...        ...   ...     ...  ..
13315                Whitefield  5 Bedroom       3453   4.0  231.00   5
13316             Richards Town      4 BHK       3600   5.0  400.00   4
13317     Raja Rajeshwari Nagar      2 BHK       1141   2.0   60.00   2
13318           Padmanabhanagar      4 BHK       4689   4.0  488.00   4
13319              Doddathoguru      1 BHK        550   1.0   17.00   1

[13246 rows x 6 columns]

In [12]:
df3['bhk'].unique()

array(['2', '4', '3', '6', '1', '8', '7', '5', '11', '9', '27', '10',
       '19', '16', '43', '14', '12', '13', '18'], dtype=object)

In [13]:
df3['total_sqft'].unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

In [14]:
def is_float(x):
    try:
        float(x)
    except:
        return False
    return True

In [15]:
wrong_format_total_sqft = df3[~df3['total_sqft'].apply(lambda x : is_float(x))]

In [16]:
wrong_format_total_sqft

location   size   total_sqft  bath    price bhk
30              Yelahanka  4 BHK  2100 - 2850   4.0  186.000   4
122                Hebbal  4 BHK  3067 - 8156   4.0  477.000   4
137    8th Phase JP Nagar  2 BHK  1042 - 1105   2.0   54.005   2
165              Sarjapur  2 BHK  1145 - 1340   2.0   43.490   2
188              KR Puram  2 BHK  1015 - 1540   2.0   56.800   2
...                   ...    ...          ...   ...      ...  ..
12975          Whitefield  2 BHK   850 - 1060   2.0   38.190   2
12990      Talaghattapura  3 BHK  1804 - 2273   3.0  122.000   3
13059              Harlur  2 BHK  1200 - 1470   2.0   72.760   2
13265               Hoodi  2 BHK  1133 - 1384   2.0   59.135   2
13299          Whitefield  4 BHK  2830 - 2882   5.0  154.500   4

[190 rows x 6 columns]

In [17]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13246 entries, 0 to 13319
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13246 non-null  object 
 1   size        13246 non-null  object 
 2   total_sqft  13246 non-null  object 
 3   bath        13246 non-null  float64
 4   price       13246 non-null  float64
 5   bhk         13246 non-null  object 
dtypes: float64(2), object(4)
memory usage: 724.4+ KB


In [18]:
def convert_total_sqft_data(x):
    try:
     return (int(x.split('-')[0]) + int(x.split('-')[1]))/2
    except:
       return x

 

In [19]:
wrong_format_total_sqft['total_sqft'] = wrong_format_total_sqft['total_sqft'].apply(lambda x : convert_total_sqft_data(x))

C:\Users\SWARAJ\AppData\Local\Temp\ipykernel_13444\3926602575.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  wrong_format_total_sqft['total_sqft'] = wrong_format_total_sqft['total_sqft'].apply(lambda x : convert_total_sqft_data(x))


In [20]:
wrong_format_total_sqft

location   size total_sqft  bath    price bhk
30              Yelahanka  4 BHK     2475.0   4.0  186.000   4
122                Hebbal  4 BHK     5611.5   4.0  477.000   4
137    8th Phase JP Nagar  2 BHK     1073.5   2.0   54.005   2
165              Sarjapur  2 BHK     1242.5   2.0   43.490   2
188              KR Puram  2 BHK     1277.5   2.0   56.800   2
...                   ...    ...        ...   ...      ...  ..
12975          Whitefield  2 BHK      955.0   2.0   38.190   2
12990      Talaghattapura  3 BHK     2038.5   3.0  122.000   3
13059              Harlur  2 BHK     1335.0   2.0   72.760   2
13265               Hoodi  2 BHK     1258.5   2.0   59.135   2
13299          Whitefield  4 BHK     2856.0   5.0  154.500   4

[190 rows x 6 columns]

In [21]:
df3['total_sqft'] = df3['total_sqft'].apply(lambda x : convert_total_sqft_data(x))

In [22]:
df3.loc[30]
# location 30

location      Yelahanka
size              4 BHK
total_sqft       2475.0
bath                4.0
price             186.0
bhk                   4
Name: 30, dtype: object

In [23]:
def sqmeter_to_num(x):
    try:
        return int(float(x.split('S')[0]))*10.7639
    except:
        return x

In [24]:
df3['total_sqft'] = df3['total_sqft'].apply(lambda x : sqmeter_to_num(x))

In [25]:
df3.loc[410]

location       Kengeri
size             1 BHK
total_sqft    365.9726
bath               1.0
price             18.5
bhk                  1
Name: 410, dtype: object

In [26]:
df4 = df3.copy()

In [27]:
df4

location       size  total_sqft  bath   price bhk
0      Electronic City Phase II      2 BHK  11366.6784   2.0   39.07   2
1              Chikka Tirupathi  4 Bedroom    27986.14   5.0  120.00   4
2                   Uttarahalli      3 BHK   15500.016   2.0   62.00   3
3            Lingadheeranahalli      3 BHK  16371.8919   3.0   95.00   3
4                      Kothanur      2 BHK    12916.68   2.0   51.00   2
...                         ...        ...         ...   ...     ...  ..
13315                Whitefield  5 Bedroom  37167.7467   4.0  231.00   5
13316             Richards Town      4 BHK    38750.04   5.0  400.00   4
13317     Raja Rajeshwari Nagar      2 BHK  12281.6099   2.0   60.00   2
13318           Padmanabhanagar      4 BHK  50471.9271   4.0  488.00   4
13319              Doddathoguru      1 BHK    5920.145   1.0   17.00   1

[13246 rows x 6 columns]

In [28]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13246 entries, 0 to 13319
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13246 non-null  object 
 1   size        13246 non-null  object 
 2   total_sqft  13246 non-null  object 
 3   bath        13246 non-null  float64
 4   price       13246 non-null  float64
 5   bhk         13246 non-null  object 
dtypes: float64(2), object(4)
memory usage: 1.2+ MB


In [29]:
df4.loc[410]

location       Kengeri
size             1 BHK
total_sqft    365.9726
bath               1.0
price             18.5
bhk                  1
Name: 410, dtype: object

In [30]:
def perch_to_sqft(x):
    try:
        return int(float(x.split('P')[0])*272.25)
    except:
        return x

In [31]:
df4['total_sqft'] = df4['total_sqft'].apply(lambda x:perch_to_sqft(x))

In [32]:
df4

location       size  total_sqft  bath   price bhk
0      Electronic City Phase II      2 BHK  11366.6784   2.0   39.07   2
1              Chikka Tirupathi  4 Bedroom    27986.14   5.0  120.00   4
2                   Uttarahalli      3 BHK   15500.016   2.0   62.00   3
3            Lingadheeranahalli      3 BHK  16371.8919   3.0   95.00   3
4                      Kothanur      2 BHK    12916.68   2.0   51.00   2
...                         ...        ...         ...   ...     ...  ..
13315                Whitefield  5 Bedroom  37167.7467   4.0  231.00   5
13316             Richards Town      4 BHK    38750.04   5.0  400.00   4
13317     Raja Rajeshwari Nagar      2 BHK  12281.6099   2.0   60.00   2
13318           Padmanabhanagar      4 BHK  50471.9271   4.0  488.00   4
13319              Doddathoguru      1 BHK    5920.145   1.0   17.00   1

[13246 rows x 6 columns]

In [33]:
def Acres_to_sqft(x):
    try:
        return int(float(x.split('A')[0])*43560)
    except:
        return x
    


In [34]:
df4['total_sqft'] = df4['total_sqft'].apply(lambda x:Acres_to_sqft(x))

In [35]:
def convert(x):
    try:
        return int(float(x.split('-')[0]) + float(x.split('-')[1]))
    except:
        return x

In [36]:
df4['total_sqft'] =df4['total_sqft'].apply(lambda x:convert(x))

In [41]:
def Cents_to_sqft(x):
    try:
        return int(x.split('C')[0])*435.56 
    except:
        return x

In [42]:
df4['total_sqft'] =df4['total_sqft'].apply(lambda x:Cents_to_sqft(x))

In [44]:
def Guntha_to_sqft(x):
    try:
     return int(x.split('G')[0])*1089
    except:
       return x

In [45]:
df4['total_sqft'] =df4['total_sqft'].apply(lambda x:Guntha_to_sqft(x))

In [46]:
pd.to_numeric(df4['total_sqft'])

0        11366.6784
1        27986.1400
2        15500.0160
3        16371.8919
4        12916.6800
            ...    
13315    37167.7467
13316    38750.0400
13317    12281.6099
13318    50471.9271
13319     5920.1450
Name: total_sqft, Length: 13246, dtype: float64

In [47]:
df5 = df4.copy()

In [48]:
df5['price_per_sqft'] = df5['price']*100000 / df5['total_sqft']

In [49]:
df5.head(2)

location       size  total_sqft  bath   price bhk  \
0  Electronic City Phase II      2 BHK  11366.6784   2.0   39.07   2   
1          Chikka Tirupathi  4 Bedroom  27986.1400   5.0  120.00   4   

   price_per_sqft  
0      343.723985  
1      428.783676

In [50]:
len(df5.location.unique())

1304

In [59]:
df5.groupby(['location']).size().sort_values(ascending=False)

location
Whitefield             534
Sarjapur  Road         392
Electronic City        302
Kanakpura Road         266
Thanisandra            233
                      ... 
 Banaswadi               1
Kanakadasa Layout        1
Kanakapur main road      1
Kanakapura  Rod          1
whitefiled               1
Length: 1304, dtype: int64

In [67]:
location_stats = df5.groupby(['location']).size().sort_values(ascending=False)

In [71]:
location_less_than_10 = location_stats[location_stats<=10]

In [72]:
df5['location'] = df5['location'].apply(lambda x: 'other' if x in location_less_than_10 else x)

In [73]:
df5

location       size  total_sqft  bath   price bhk  \
0      Electronic City Phase II      2 BHK  11366.6784   2.0   39.07   2   
1              Chikka Tirupathi  4 Bedroom  27986.1400   5.0  120.00   4   
2                   Uttarahalli      3 BHK  15500.0160   2.0   62.00   3   
3            Lingadheeranahalli      3 BHK  16371.8919   3.0   95.00   3   
4                      Kothanur      2 BHK  12916.6800   2.0   51.00   2   
...                         ...        ...         ...   ...     ...  ..   
13315                Whitefield  5 Bedroom  37167.7467   4.0  231.00   5   
13316                     other      4 BHK  38750.0400   5.0  400.00   4   
13317     Raja Rajeshwari Nagar      2 BHK  12281.6099   2.0   60.00   2   
13318           Padmanabhanagar      4 BHK  50471.9271   4.0  488.00   4   
13319              Doddathoguru      1 BHK   5920.1450   1.0   17.00   1   

       price_per_sqft  
0          343.723985  
1          428.783676  
2          399.999587  
3          580.262810  
4          394.838302  
...               ...  
13315      621.506603  
13316     1032.256999  
13317      488.535302  
13318      966.874118  
13319      287.155129  

[13246 rows x 7 columns]

In [76]:
len(df5.location.unique())

242

on an average each bedroom occupies 300sqft or more , we will check that

In [80]:
df5.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13246 entries, 0 to 13319
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   location        13246 non-null  object 
 1   size            13246 non-null  object 
 2   total_sqft      13246 non-null  float64
 3   bath            13246 non-null  float64
 4   price           13246 non-null  float64
 5   bhk             13246 non-null  object 
 6   price_per_sqft  13246 non-null  float64
dtypes: float64(4), object(3)
memory usage: 1.3+ MB


In [83]:
df5['bhk'] = pd.to_numeric(df5['bhk'])

In [84]:
df5.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13246 entries, 0 to 13319
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   location        13246 non-null  object 
 1   size            13246 non-null  object 
 2   total_sqft      13246 non-null  float64
 3   bath            13246 non-null  float64
 4   price           13246 non-null  float64
 5   bhk             13246 non-null  int64  
 6   price_per_sqft  13246 non-null  float64
dtypes: float64(4), int64(1), object(2)
memory usage: 1.3+ MB


In [86]:
df5[df5['total_sqft']/df5['bhk']<300]

location       size  total_sqft  bath  price  bhk  \
349              other  3 Bedroom    118.4029   3.0   74.0    3   
1020             other      1 BHK    161.4585   1.0   30.0    1   
1122             other  5 Bedroom    258.3336   2.0  150.0    5   
4086    Sarjapur  Road  4 Bedroom     10.7639   4.0  120.0    4   
4972             other      7 BHK     53.8195   7.0  115.0    7   
5453             other  6 Bedroom   1306.6800   6.0   75.0    6   
5668   Judicial Layout  5 Bedroom    484.3755   6.0  400.0    5   
11558       Whitefield  4 Bedroom    645.8340   4.0  218.0    4   
12252            other      5 BHK    258.3336   5.0   75.0    5   

       price_per_sqft  
349      6.249847e+04  
1020     1.858063e+04  
1122     5.806446e+04  
4086     1.114838e+06  
4972     2.136772e+05  
5453     5.739737e+03  
5668     8.258056e+04  
11558    3.375480e+04  
12252    2.903223e+04

These are errors or outliers

In [87]:
df6 = df5[~(df5['total_sqft']/df5['bhk']<300)]

In [91]:
df5.shape


(13246, 7)

In [90]:

df6.shape

(13237, 7)

In [56]:
df5['location'].apply(lambda x: x.strip())
# to remove extra spaces before and after the words

0        Electronic City Phase II
1                Chikka Tirupathi
2                     Uttarahalli
3              Lingadheeranahalli
4                        Kothanur
                   ...           
13315                  Whitefield
13316               Richards Town
13317       Raja Rajeshwari Nagar
13318             Padmanabhanagar
13319                Doddathoguru
Name: location, Length: 13246, dtype: object